### PR TITLE
Implement redis

### PR DESCRIPTION
**Pull request by ajax-yurkovskyi-t**

### CODE REVIEW

File: iot-management-device/src/main/kotlin/com/example/iotmanagementdevice/repository/RedisDeviceRepository.kt [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+        val key = createDeviceKey(deviceId)
+        return reactiveRedisTemplate.opsForValue().get(key)
+            .handle { item, sink ->
+                if (item.isEmpty()) {
+                    sink.error(EntityNotFoundException("Device with id $deviceId not found"))
+                } else {
+                    sink.next(item)
+                }
+            }
+            .map { mapper.readValue<MongoDevice>(it) }
```
Comment (ajax-marchenko-s) [inline iot-management-device/src/main/kotlin/com/example/iotmanagementdevice/repository/RedisDeviceRepository.kt:48]: what if parsing fails?

File: iot-management-device/src/main/kotlin/com/example/iotmanagementdevice/repository/RedisDeviceRepository.kt [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        ).thenReturn(device)
+    }
+
+    private fun saveDeviceWithRetries(device: MongoDevice) {
+        saveDeviceToRedis(device).retryWhen(
+            Retry.backoff(MAX_RETRIES, Duration.ofSeconds(RETRY_BACKOFF_SECONDS))
+                .filter { error -> isErrorFromRedis(error) }
+                .doBeforeRetry { retrySignal ->
+                    log.warn(
+                        "Retrying saving device to Redis because of: ${retrySignal.failure().message}"
```
Comment (ajax-marchenko-s) [inline iot-management-device/src/main/kotlin/com/example/iotmanagementdevice/repository/RedisDeviceRepository.kt:83]: fix building string

**Review by ajax-marchenko-s — COMMENTED** [review]
Great work!

File: iot-management-device/src/main/kotlin/com/example/iotmanagementdevice/repository/RedisDeviceRepository.kt [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+import reactor.kotlin.core.publisher.switchIfEmpty
+import reactor.util.retry.Retry
+import java.net.SocketException
+import java.time.Duration
+
+@Repository
+@Primary
+class RedisDeviceRepository(
+    @Value("\${spring.data.redis.ttl.minutes}")
+    private val redisExpirationTimeoutInMinutes: Long,
```
Comment (ajax-shulepa-a) [inline iot-management-device/src/main/kotlin/com/example/iotmanagementdevice/repository/RedisDeviceRepository.kt:26]: can it be duration?

File: iot-management-device/src/main/kotlin/com/example/iotmanagementdevice/repository/RedisDeviceRepository.kt [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    private val mongoDeviceRepository: MongoDeviceRepository,
+    private val reactiveRedisTemplate: ReactiveRedisTemplate<String, ByteArray>,
+    private val mapper: ObjectMapper,
+) : DeviceRepository by mongoDeviceRepository {
+
+    override fun save(device: MongoDevice): Mono<MongoDevice> {
+        return mongoDeviceRepository.save(device).doOnSuccess {
```
Comment (ajax-shulepa-a) [inline iot-management-device/src/main/kotlin/com/example/iotmanagementdevice/repository/RedisDeviceRepository.kt:33]:
```suggestion
        return mongoDeviceRepository.save(device)
            .doOnSuccess {
```

File: iot-management-device/src/main/kotlin/com/example/iotmanagementdevice/repository/RedisDeviceRepository.kt [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+        }
+            .switchIfEmpty {
+                reactiveRedisTemplate.opsForValue()
+                    .set(key, byteArrayOf(), Duration.ofMinutes(redisExpirationTimeoutInMinutes))
+                    .then(Mono.empty())
+            }
+    }
+
+    private fun isErrorFromRedis(throwable: Throwable): Boolean {
+        return redisErrors.any { it.isInstance(throwable) }
```
Comment (ajax-shulepa-a) [inline iot-management-device/src/main/kotlin/com/example/iotmanagementdevice/repository/RedisDeviceRepository.kt:109]: what's the complexity here?

File: iot-management-device/src/main/kotlin/com/example/iotmanagementdevice/repository/RedisDeviceRepository.kt [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+
+        private val redisErrors = setOf(
+            RedisConnectionFailureException::class,
+            RedisException::class,
+            SocketException::class,
+            QueryTimeoutException::class,
+        )
+
+        fun createDeviceKey(key: String): String {
+            return "$KEY_PREFIX$key"
```
Comment (ajax-shulepa-a) [inline iot-management-device/src/main/kotlin/com/example/iotmanagementdevice/repository/RedisDeviceRepository.kt:126]:
```suggestion
            return "$KEY_PREFIX::$key"
```
it's better to separate prefix and id

File: iot-management-device/src/main/kotlin/com/example/iotmanagementdevice/repository/RedisDeviceRepository.kt [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+    override fun deleteById(deviceId: String): Mono<Unit> {
+        val key = createDeviceKey(deviceId)
+        return mongoDeviceRepository.deleteById(deviceId)
+            .flatMap {
+                reactiveRedisTemplate.unlink(key)
+                    .onErrorResume(::isErrorFromRedis) {
+                        log.warn("Redis failed to remove user from cache", it)
+                        Mono.empty()
+                    }.thenReturn(Unit)
+            }.thenReturn(Unit)
```
Comment (ajax-shulepa-a) [inline iot-management-device/src/main/kotlin/com/example/iotmanagementdevice/repository/RedisDeviceRepository.kt:68]: why do you need thenReturn(Unit) here?

File: iot-management-device/src/main/kotlin/com/example/iotmanagementdevice/repository/RedisDeviceRepository.kt [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+                    )
+                }
+        )
+            .subscribeOn(Schedulers.boundedElastic())
+            .subscribe()
+    }
+
+    private fun findInMongoAndWriteToRedis(deviceId: String): Mono<MongoDevice> {
+        val key = createDeviceKey(deviceId)
+        return mongoDeviceRepository.findById(deviceId).flatMap {
```
Comment (ajax-shulepa-a) [inline iot-management-device/src/main/kotlin/com/example/iotmanagementdevice/repository/RedisDeviceRepository.kt:98]: I see that sometimes a reactor operator is located on the same line, and sometimes it's placed on the new line

be consistent

File: iot-management-device/src/main/kotlin/com/example/iotmanagementdevice/repository/RedisDeviceRepository.kt (hunk 5 above)
Comment (ajax-yurkovskyi-t) [reply iot-management-device/src/main/kotlin/com/example/iotmanagementdevice/repository/RedisDeviceRepository.kt]: O(n), change to O(1)

File: iot-management-device/src/main/kotlin/com/example/iotmanagementdevice/repository/RedisDeviceRepository.kt (hunk 7 above)
Comment (ajax-yurkovskyi-t) [reply iot-management-device/src/main/kotlin/com/example/iotmanagementdevice/repository/RedisDeviceRepository.kt]: remove redundant one

File: iot-management-device/src/main/kotlin/com/example/iotmanagementdevice/repository/RedisDeviceRepository.kt [hunk 9]
```diff
@@ -95,18 +100,23 @@ class RedisDeviceRepository(
 
     private fun findInMongoAndWriteToRedis(deviceId: String): Mono<MongoDevice> {
         val key = createDeviceKey(deviceId)
-        return mongoDeviceRepository.findById(deviceId).flatMap {
-            saveDeviceToRedis(it)
-        }
+        return mongoDeviceRepository.findById(deviceId)
+            .flatMap {
+                saveDeviceToRedis(it)
+            }
```
Comment (ajax-shulepa-a) [inline iot-management-device/src/main/kotlin/com/example/iotmanagementdevice/repository/RedisDeviceRepository.kt:106]:
```suggestion
            .flatMap { saveDeviceToRedis(it) }
```

**Review by ajax-shulepa-a — APPROVED** [review]

**Pull request merged**